components/base/sensorcontrolled/sensorcontrolled.go

```diff
@@ -158,11 +171,47 @@ func (sb *sensorBase) Reconfigure(ctx context.Context, deps resource.Dependencie
 		return errors.Wrapf(err, "no base named (%s)", newConf.Base)
 	}
 
-	if sb.velocities != nil && useControlLoop {
-		sb.controlLoopConfig = sb.createControlLoopConfig()
-		if err := sb.setupControlLoops(); err != nil {
-			return err
+	if sb.velocities != nil && len(newConf.ControlParameters) != 0 {
+		// assign linear and angular PID correctly based on the given type
+		var linear, angular basePIDConfig
+		for _, c := range newConf.ControlParameters {
+			if c.Type == typeLinVel {
+				linear = c
+			} else if c.Type == typeAngVel {
+				angular = c
+			}
 		}
+
+		sb.mu.Unlock()
+		// check if both linear and angular need to be tuned, and if so start by tuning linear
+		if !(linear.P == 0.0 && linear.I == 0.0 && linear.D == 0.0 &&
+			angular.P == 0.0 && angular.I == 0.0 && angular.D == 0.0) {
+			sb.controlLoopConfig = sb.createControlLoopConfig(linear, angular)
+			if err := sb.setupControlLoops(); err != nil {
+				sb.mu.Lock()
+				return err
+			}
+		} else {
+			// to tune linear PID values, angular PI values must be non-zero
+			fakeConf := basePIDConfig{Type: typeAngVel, P: 0.5, I: 0.5, D: 0.0}
+			sb.logger.Info("tuning linear PID")
+			if err := sb.autoTuningProcess(ctx, linear, fakeConf); err != nil {
+				sb.mu.Lock()
+				return err
+			}
+			if err := sb.Stop(ctx, nil); err != nil {
+				sb.mu.Lock()
+				return err
+			}
+			// to tune angular PID values, linear PI values must be non-zero
+			fakeConf.Type = typeLinVel
+			sb.logger.Info("tuning angular PID")
+			if err := sb.autoTuningProcess(ctx, fakeConf, angular); err != nil {
+				sb.mu.Lock()
+				return err
+			}
+		}
+		sb.mu.Lock()
```

Comment:
This lock feels irrelevant.

Reply:
Thanks for adding the comments ❤️ This makes reading this function so much easier!!